torch/onnx/_type_utils.py

```diff
@@ -0,0 +1,245 @@
+        return _SCALAR_TYPE_TO_DTYPE[self]
+
+    def onnx_type(self) -> _C_onnx.TensorProtoDataType:
+        """Convert a ScalarType to an ONNX data type."""
+        if not self.onnx_compatible():
+            raise ValueError(f"Scalar type {self} is not compatible with ONNX")
+        return _SCALAR_TYPE_TO_ONNX[self]
+
+    def onnx_compatible(self) -> bool:
+        """Return whether this ScalarType is compatible with ONNX."""
+        return self in _ONNX_COMPATIBLE_TYPES
+
+
+def valid_scalar_name(scalar_name: Union[ScalarName, str]) -> bool:
+    """Return whether the given scalar name is a valid JIT scalar type name."""
+    return scalar_name in _SCALAR_NAME_TO_TYPE
+
+
+def valid_torch_name(torch_name: Union[TorchName, str]) -> bool:
+    """Return whether the given torch name is a valid torch type name."""
+    return torch_name in _TORCH_NAME_TO_SCALAR_TYPE
+
+
+# https://github.com/pytorch/pytorch/blob/344defc9733a45fee8d0c4d3f5530f631e823196/c10/core/ScalarType.h
+_SCALAR_TYPE_TO_NAME: Dict[ScalarType, ScalarName] = {
+    ScalarType.BOOL: "Bool",
+    ScalarType.UINT8: "Byte",
+    ScalarType.INT8: "Char",
+    ScalarType.INT16: "Short",
+    ScalarType.INT: "Int",
+    ScalarType.INT64: "Long",
+    ScalarType.HALF: "Half",
+    ScalarType.FLOAT: "Float",
+    ScalarType.DOUBLE: "Double",
+    ScalarType.COMPLEX32: "ComplexHalf",
+    ScalarType.COMPLEX64: "ComplexFloat",
+    ScalarType.COMPLEX128: "ComplexDouble",
+    ScalarType.QINT8: "QInt8",
+    ScalarType.QUINT8: "QUInt8",
+    ScalarType.QINT32: "QInt32",
+    ScalarType.BFLOAT16: "BFloat16",
+    ScalarType.UNDEFINED: "Undefined",
+}
+
+_SCALAR_NAME_TO_TYPE: Dict[ScalarName, ScalarType] = {
+    v: k for k, v in _SCALAR_TYPE_TO_NAME.items()
+}
+
+_SCALAR_TYPE_TO_TORCH_NAME: Dict[ScalarType, TorchName] = {
+    ScalarType.BOOL: "bool",
+    ScalarType.UINT8: "uint8_t",
+    ScalarType.INT8: "int8_t",
+    ScalarType.INT16: "int16_t",
+    ScalarType.INT: "int",
+    ScalarType.INT64: "int64_t",
+    ScalarType.HALF: "half",
+    ScalarType.FLOAT: "float",
+    ScalarType.DOUBLE: "double",
+    ScalarType.COMPLEX32: "complex32",
+    ScalarType.COMPLEX64: "complex64",
+    ScalarType.COMPLEX128: "complex128",
+    ScalarType.QINT8: "qint8",
+    ScalarType.QUINT8: "quint8",
+    ScalarType.QINT32: "qint32",
+    ScalarType.BFLOAT16: "bfloat16",
+}
+
+_TORCH_NAME_TO_SCALAR_TYPE: Dict[TorchName, ScalarType] = {
+    v: k for k, v in _SCALAR_TYPE_TO_TORCH_NAME.items()
+}
+
+_ONNX_COMPATIBLE_TYPES = frozenset(
+    (
+        ScalarType.BOOL,
+        ScalarType.UINT8,
+        ScalarType.INT8,
+        ScalarType.INT16,
+        ScalarType.INT,
+        ScalarType.INT64,
+        ScalarType.HALF,
+        ScalarType.FLOAT,
+        ScalarType.DOUBLE,
+        ScalarType.COMPLEX64,
+        ScalarType.COMPLEX128,
+        ScalarType.BFLOAT16,
+        ScalarType.QINT8,
+        ScalarType.QUINT8,
+        ScalarType.QINT32,
+    )
+)
+
+_SCALAR_TYPE_TO_ONNX = {
```

Comment:
What's the relationship between `_SCALAR_TYPE_TO_ONNX ` and `_ONNX_COMPATIBLE_TYPES`?

Reply:
Types like `ScalarType.COMPLEX32` are previously mapped to `_C_onnx.TensorProtoDataType.UNDEFINED`. Do you think we should drop this mapping instead?

Reply:
Is it correct that `_SCALAR_TYPE_TO_ONNX[ScalarType.COMPLEX32]` is not reachable based on current api `onnx_type`?

Reply:
If that's the case I feel these two can be merged.

Reply:
I removed the set and checked additionally UNDEFINED and COMPLEX32 in `onnx_compatible()`